From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp100: Export tr1::_Xfunc, _Mutex(_Uninitialized), _GetCurrentThreadId and the unsigned short reader.

msvcp100: tr1::_Xfunc (throws std::bad_function_call, as
_Xbad_function_call does in the later libraries), _Mutex(_Uninitialized)
(leaves the object as it is; the static mutexes are set up by
_Mutex_ctor), details::_GetCurrentThreadId as in msvcp110, and the
unsigned short operator>>(unsigned short &) and time_get::get with a
format, as msvcp110 has them.
---
 dlls/msvcp100/msvcp100.spec | 20 ++++++++++----------
 dlls/msvcp90/misc.c         | 15 +++++++++++++++
 2 files changed, 25 insertions(+), 10 deletions(-)

diff --git a/dlls/msvcp100/msvcp100.spec b/dlls/msvcp100/msvcp100.spec
index 23017f2..c5ad8ee 100644
--- a/dlls/msvcp100/msvcp100.spec
+++ b/dlls/msvcp100/msvcp100.spec
@@ -293,8 +293,8 @@
 @ thiscall -arch=i386 ??0_Lockit@std@@QAE@XZ(ptr) _Lockit_ctor
 @ cdecl -arch=win64 ??0_Lockit@std@@QEAA@XZ(ptr) _Lockit_ctor
 @ stub -arch=arm ??0_Mutex@std@@QAA@W4_Uninitialized@1@@Z
-@ stub -arch=i386 ??0_Mutex@std@@QAE@W4_Uninitialized@1@@Z
-@ stub -arch=win64 ??0_Mutex@std@@QEAA@W4_Uninitialized@1@@Z
+@ thiscall -arch=i386 ??0_Mutex@std@@QAE@W4_Uninitialized@1@@Z(ptr long) mutex_ctor_uninitialized
+@ cdecl -arch=win64 ??0_Mutex@std@@QEAA@W4_Uninitialized@1@@Z(ptr long) mutex_ctor_uninitialized
 @ cdecl -arch=arm ??0_Mutex@std@@QAA@XZ(ptr) mutex_ctor
 @ thiscall -arch=i386 ??0_Mutex@std@@QAE@XZ(ptr) mutex_ctor
 @ cdecl -arch=win64 ??0_Mutex@std@@QEAA@XZ(ptr) mutex_ctor
@@ -685,9 +685,9 @@
 @ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAF@Z(ptr ptr) basic_istream_short_read_short
 @ thiscall -arch=i386 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV01@AAF@Z(ptr ptr) basic_istream_short_read_short
 @ cdecl -arch=win64 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEAF@Z(ptr ptr) basic_istream_short_read_short
-@ stub -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAG@Z
-@ stub -arch=i386 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV01@AAG@Z
-@ stub -arch=win64 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEAG@Z
+@ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAG@Z(ptr ptr) basic_istream_short_read_ushort
+@ thiscall -arch=i386 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV01@AAG@Z(ptr ptr) basic_istream_short_read_ushort
+@ cdecl -arch=win64 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEAG@Z(ptr ptr) basic_istream_short_read_ushort
 @ cdecl -arch=arm ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV01@AAH@Z(ptr ptr) basic_istream_short_read_int
 @ thiscall -arch=i386 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV01@AAH@Z(ptr ptr) basic_istream_short_read_int
 @ cdecl -arch=win64 ??5?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV01@AEAH@Z(ptr ptr) basic_istream_short_read_int
@@ -1172,7 +1172,7 @@
 @ cdecl -arch=win64 ?_Fput@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@AEBA?AV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@2@V32@AEAVios_base@2@_WPEBD_K444@Z(ptr ptr ptr ptr long str long long long long) num_put_wchar__Fput
 @ cdecl -arch=win32 ?_GetCombinableSize@details@Concurrency@@YAIXZ() _GetCombinableSize
 @ cdecl -arch=win64 ?_GetCombinableSize@details@Concurrency@@YA_KXZ() _GetCombinableSize
-@ stub ?_GetCurrentThreadId@details@Concurrency@@YAKXZ
+@ cdecl ?_GetCurrentThreadId@details@Concurrency@@YAKXZ() _GetCurrentThreadId
 @ cdecl -arch=win32 ?_Getcat@?$codecvt@DDH@std@@SAIPAPBVfacet@locale@2@PBV42@@Z(ptr ptr) codecvt_char__Getcat
 @ cdecl -arch=win64 ?_Getcat@?$codecvt@DDH@std@@SA_KPEAPEBVfacet@locale@2@PEBV42@@Z(ptr ptr) codecvt_char__Getcat
 @ cdecl -arch=win32 ?_Getcat@?$codecvt@GDH@std@@SAIPAPBVfacet@locale@2@PBV42@@Z(ptr ptr) codecvt_short__Getcat
@@ -1697,7 +1697,7 @@
 @ cdecl ?_XLgamma@tr1@std@@YANN@Z(double) std__XLgamma_double
 @ cdecl ?_XLgamma@tr1@std@@YAOO@Z(double) std__XLgamma_double
 @ stub ?_Xbad@tr1@std@@YAXW4error_type@regex_constants@12@@Z
-@ stub ?_Xfunc@tr1@std@@YAXXZ
+@ cdecl ?_Xfunc@tr1@std@@YAXXZ() _Xbad_function_call
 @ cdecl -arch=win32 ?_Xinvalid_argument@std@@YAXPBD@Z(str) _Xinvalid_argument
 @ cdecl -arch=win64 ?_Xinvalid_argument@std@@YAXPEBD@Z(str) _Xinvalid_argument
 @ cdecl -arch=win32 ?_Xlength_error@std@@YAXPBD@Z(str) _Xlength_error
@@ -2524,9 +2524,9 @@
 @ cdecl -arch=arm ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@DD@Z(ptr ptr ptr long ptr long ptr ptr ptr long long) time_get_wchar_get
 @ thiscall -arch=i386 ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBE?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@DD@Z(ptr ptr ptr long ptr long ptr ptr ptr long long) time_get_wchar_get
 @ cdecl -arch=win64 ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QEBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AEAVios_base@2@AEAHPEAUtm@@DD@Z(ptr ptr ptr long ptr long ptr ptr ptr long long) time_get_wchar_get
-@ stub -arch=arm ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@PBG4@Z
-@ stub -arch=i386 ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBE?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@PBG4@Z
-@ stub -arch=win64 ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QEBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AEAVios_base@2@AEAHPEAUtm@@PEBG4@Z
+@ cdecl -arch=arm ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@PBG4@Z(ptr ptr ptr long ptr long ptr ptr ptr ptr ptr) time_get_wchar_get_fmt
+@ thiscall -arch=i386 ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBE?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@PBG4@Z(ptr ptr ptr long ptr long ptr ptr ptr ptr ptr) time_get_wchar_get_fmt
+@ cdecl -arch=win64 ?get@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QEBA?AV?$istreambuf_iterator@GU?$char_traits@G@std@@@2@V32@0AEAVios_base@2@AEAHPEAUtm@@PEBG4@Z(ptr ptr ptr ptr ptr ptr ptr ptr ptr) time_get_wchar_get_fmt
 @ cdecl -arch=arm ?get@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@QBA?AV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@DD@Z(ptr ptr ptr long ptr long ptr ptr ptr long long) time_get_wchar_get
 @ thiscall -arch=i386 ?get@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@QBE?AV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@2@V32@0AAVios_base@2@AAHPAUtm@@DD@Z(ptr ptr ptr long ptr long ptr ptr ptr long long) time_get_wchar_get
 @ cdecl -arch=win64 ?get@?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@QEBA?AV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@2@V32@0AEAVios_base@2@AEAHPEAUtm@@DD@Z(ptr ptr ptr long ptr long ptr ptr ptr long long) time_get_wchar_get
diff --git a/dlls/msvcp90/misc.c b/dlls/msvcp90/misc.c
index 2075eae..701c781 100644
--- a/dlls/msvcp90/misc.c
+++ b/dlls/msvcp90/misc.c
@@ -273,6 +273,18 @@ mutex* __thiscall mutex_ctor(mutex *this)
     return this;
 }
 
+#if _MSVCP_VER == 100
+/* ??0_Mutex@std@@QAE@W4_Uninitialized@1@@Z */
+/* ??0_Mutex@std@@QEAA@W4_Uninitialized@1@@Z */
+/* leaves the mutex as it is: the static ones are set up by _Mutex_ctor */
+DEFINE_THISCALL_WRAPPER(mutex_ctor_uninitialized, 8)
+mutex* __thiscall mutex_ctor_uninitialized(mutex *this, int uninitialized)
+{
+    TRACE("(%p %d)\n", this, uninitialized);
+    return this;
+}
+#endif
+
 /* ??1_Mutex@std@@QAE@XZ */
 /* ??1_Mutex@std@@QEAA@XZ */
 DEFINE_THISCALL_WRAPPER(mutex_dtor, 4)
@@ -2275,6 +2287,9 @@ void __thiscall basic_ios__Add_vtordisp(void *this)
     TRACE("(%p)\n", this);
 }
 
+#endif
+
+#if _MSVCP_VER >= 100 && _MSVCP_VER <= 110
 /* ?_GetCurrentThreadId@details@Concurrency@@YAKXZ */
 DWORD __cdecl _GetCurrentThreadId(void)
 {
